nilearn/glm/first_level/first_level.py

```diff
@@ -959,107 +950,387 @@ def first_level_from_bids(dataset_path, task_label, space_label=None,
+                           supported_filters: list[str] | None = None,
+                           extra_filter: list[Tuple[str, str]] | None = None
+                           ) -> list[Tuple[str, str]]:
+    """Return a filter to specific files from a BIDS dataset.
+
+    Parameters
+    ----------
+    task_label : :obj:`str`
+        Task_label as specified in the file names like _task-<task_label>_.
+
+    space_label : :obj:`str` or None, optional
+        Specifies the space label of the preprocessed bold.nii images.
+        As they are specified in the file names like _space-<space_label>_.
+
+    supported_filters : :obj:`list` of :obj:`str` or None, optional
+        List of authorized BIDS entities
+
+    extra_filter : :obj:`list` of :obj:`tuple` (str, str) or None, optional
+        _description_
+
+    Returns
+    -------
+    Filter to be used by :func:`get_bids_files`: \
+        :obj:`list` of :obj:`tuple` (str, str)
+        filters
+
+    """
+    filters = [('task', task_label)]
+
+    if space_label is not None:
+        filters.append(('space', space_label))
+
+    if extra_filter and supported_filters:
+        for this_filter in extra_filter:
+
+            if this_filter[0] not in supported_filters:
+                warn(f"The filter {this_filter} will be skipped. "
+                     f"'{this_filter[0]}' is not among the supported filters. "
+                     f"Only {supported_filters} are allowed.")
+                continue
+
+            filters.append(this_filter)
+
+    return filters
+
+
+def _check_bids_image_list(imgs: list[str] | None,
+                           sub_label: str,
+                           filters: list[Tuple[str, str]]) -> None:
+    """Check input BIDS images.
+
+    Check that:
+        - some images were found
+        - if more than one image was found, check that there is not more than
+          one image for a given session / run combination.
+
+    Parameters
+    ----------
+    imgs : list[str] or None
+        List of image fullpath filenames.
+
+    sub_label : :obj:`str`
+        Subject label as specified in the file names like _sub-<sub_label>_.
+
+    filters : :obj:`list` of :obj:`tuple` (str, str)
+        Filters of the form (field, label) used to select the files.
+        See :func:`get_bids_files`.
+
+    """
+    if not imgs:
+        raise ValueError('No bold files found '
+                         f'for subject {sub_label} '
+                         f'for filter: {filters}')
+
+    if len(imgs) <= 1:
+        return
+
+    # If there is more than one file for the same (ses, run),
+    # likely we have an issue of underspecification of filters.
+    # If more than one run is present the run field is mandatory in BIDS
+    # as well as the ses field if more than one session is present.
+    msg_end = ('Please verify that the desc_label and space_label labels ' +
+               'corresponding to the BIDS spec were correctly specified.')
```

Comment:
done

Reply:
> I had commented on this in a previous review, to include the comment's content in the message since it pinpoints the problem & indicates the fix. If you disagree & want to keep it as is, I won't try to block anything, pointing it again, in case you had simply overlooked it in this location.

Not sure you saw it but the message would now read:

```python
   msg_start = ("More than one nifti image found\n "
                 f"for subject: '{sub_label}'\n"
                 f"for filters: {filters}\n")
   msg_end = ("Adapt that the arguments"
               "'task_label', 'space_label' and 'img_filters'"
               "to make sure only file per session, per run gets selected.")
```

Reply:
> Not sure you saw it

Hehe, clearly I didn't see it 😁 .
Did you mean `"Adapt"` -> `"Ensure"`
```python
msg_end = ("Ensure that the arguments"
```
and
```python
"to make sure only one file per session and per run
```

However, I am not certain the message explains what changes to make to adapt/ensure these args only select one field. Could be because I don't use the functionality any longer. If I am incorrect, then ignore this.

Reply:
("Too many images fit your query. Please specify it further by setting e.g. some required  task labels, space label or img_filters") ?

Reply:
Sounds good to me.
Went with that.